src/models/tests/pairwise_union_reach_surface_test.py

```diff
@@ -0,0 +1,125 @@
+    self._max_reach = self._data[0].reach()
+
+  def by_impressions(self,
+                     impressions: [int],
+                     max_frequency: int = 1) -> ReachPoint:
+    """Returns the estimated reach for a given impression vector."""
+    kplus_frequencies = [
+        min(sum(impressions), self.max_reach) // i
+        for i in range(1, max_frequency + 1)
+    ]
+    return ReachPoint(impressions, kplus_frequencies)
+
+
+class PairwiseUnionReachSurfaceTest(absltest.TestCase):
+
+  def assertPointsAlmostEqualToPrediction(self,
+                                          surface,
+                                          reach_points,
+                                          tolerance=0.0001):
+    for reach_point in reach_points:
+      self.assertAlmostEqual(
+          surface.by_impressions(reach_point.impressions).reach(),
+          reach_point.reach(),
+          delta=reach_point.reach() * tolerance)
+
+  def generate_true_reach(self, a, reach_curves, impressions):
+    p = len(reach_curves)
+    reach_vector = [
+        reach_curve.by_impressions(impression).reach()
+        for reach_curve, impression in zip(reach_curves, impressions)
+    ]
+    reach = sum(reach_vector)
+    for i in range(len(reach_curves)):
+      for j in range(len(reach_curves)):
+        reach -= (a[i * p + j] * reach_vector[i] * reach_vector[j]) / (
+            max(reach_curves[i].max_reach, reach_curves[j].max_reach) * 2)
+    return ReachPoint(impressions, [reach])
+
+  def generate_sample_reach_curves(self, num_publishers, decay_rate,
+                                   universe_size):
+    max_reaches = [
+        universe_size * (decay_rate**pub_num)
+        for pub_num in range(num_publishers)
+    ]
+    return [
+        LinearCappedReachCurve([ReachPoint([max_reach], (max_reach,))])
+        for max_reach in max_reaches
+    ]
+
+  def generate_sample_matrix_a(self, num_publishers):
+    true_a = np.array([
+        np.random.dirichlet(np.ones(num_publishers)) * np.random.uniform()
+        for _ in range(num_publishers)
+    ]).flatten()
+
+    # satisfy the constraint a_i_i = 0
+    for i in range(num_publishers):
+      true_a[i * num_publishers + i] = 0
+
+    # satisfy the constraint a_i_j = a_j_i by copying top half to bottom half.
+    for i in range(num_publishers):
+      for j in range(i+1, num_publishers):
+        true_a[j * num_publishers + i] = true_a[i * num_publishers + j]
+    return true_a
+
+  def generate_sample_reach_points(self, true_a, reach_curves, size,
+                                   universe_size):
+    reach_points = []
+    random_generator = np.random.default_rng(1)
+    for _ in range(size):
+      impressions = [[random_generator.uniform(0, universe_size / 2)]
+                     for _ in range(len(reach_curves))]
+      reach_points.append(
+          self.generate_true_reach(true_a, reach_curves, impressions))
+
+    return reach_points
+
+  def test_by_impressions(self):
```

Comment:
All of this looks good.  I have one additional request, if you don't mind.  Could you please add another two unit tests:

(a) Two reach curves, A and B, with no overlap.  If r_A(i) is the reach curve for A and r_B(i) is the reach curve for B, then the pairwise union model should satisfy r(i_A, i_B) = r_A(i_A) + r_B(i_B).

(b) Two reach curves, A and B, where the viewers in each curve are chosen uniformly at random from a universe of size N.  In this case, the pairwise union model should satisfy r(i_A, i_B) = r_A(i_A) + r_B(i_B) - r_A(i_A) * r_B(i_B) / N.

Thanks for coding all of this.  It looks great!

Reply:
thank you Matthew, I added those test cases too. Hope these make sense.